test/integration/filters/sleeping_filter.cc

```diff
@@ -0,0 +1,62 @@
+#include <string>
+
+#include "envoy/registry/registry.h"
+
+#include "extensions/filters/http/common/factory_base.h"
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+#include "test/integration/filters/sleeping_filter.pb.h"
+#include "test/integration/filters/sleeping_filter.pb.validate.h"
+#include "test/test_common/test_time_system.h"
+
+namespace Envoy {
+
+// A test filter that inserts sleeps during the decode/encode phase. Sleep is induced by invoking a
+// thread sleep, and it is done so to simulate a slow/hanging filter that causes the thread
+// executing the filter callbacks to block.
+class SleepingFilter : public Http::PassThroughFilter {
+public:
+  // The dynamic_cast from Event::TimeSystem& to Event::TestTimeSystem& is necessary for exposing
+  // the sleep API to this filter.
+  SleepingFilter(Event::TimeSystem& time_system, uint64_t sleep_ms)
+      : PassThroughFilter(), time_system_(dynamic_cast<Event::TestTimeSystem&>(time_system)),
+        sleep_ms_(sleep_ms) {}
+
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool) {
+    time_system_.sleep(std::chrono::milliseconds(sleep_ms_));
+    return Http::FilterDataStatus::Continue;
+  }
+
+  Http::FilterDataStatus encodeData(Buffer::Instance&, bool) {
+    time_system_.sleep(std::chrono::milliseconds(sleep_ms_));
+    return Http::FilterDataStatus::Continue;
+  }
+
+private:
+  Event::TestTimeSystem& time_system_;
+  uint64_t sleep_ms_;
```

Comment:
Use Event::TimeSystem::Duration for the type of this variable, and throughout the call-stack, rather than uint64_t. Then you don't need units in the variable-name, or manual conversion from nanos.